sheepdog/auth/__init__.py

```diff
@@ -113,9 +115,27 @@ def authorize_and_call(*args, **kwargs):
 def authorize(program, project, roles):
     resource = "/programs/{}/projects/{}".format(program, project)
     jwt = get_jwt_from_header()
-    authz = flask.current_app.auth.auth_request(
-        jwt=jwt, service="sheepdog", methods=roles, resources=[resource]
-    )
+    cache_key = str(hash((jwt, "sheepdog", tuple(roles), (resource))))
+    authz = None
+    try:
+        if AUTHZ_CACHE.has(cache_key):
+            authz = AUTHZ_CACHE.get(cache_key)
+        else:
+            authz = flask.current_app.auth.auth_request(
+                jwt=jwt, service="sheepdog", methods=roles, resources=[resource]
+            )
+            logger.info(
+                f"Retrieveing response from arborist: {authz} with {type(authz)=}"
+            )
+            AUTHZ_CACHE.set(cache_key, authz)
+    # The caching library raises an UnboundLocalError during unit tests due to a known bug.
+    # This workaround prevents the error from occurring in test environments.
+    except UnboundLocalError as e:
+        logger.error("Catching error caused by caching library: {}".format(e))
+        authz = flask.current_app.auth.auth_request(
+            jwt=jwt, service="sheepdog", methods=roles, resources=[resource]
+        )
```

Comment:
any way we could move this to the unit tests, then?
and please add a link to the known bug - github issue or such

Reply:
I’ve removed the try-except block entirely and instead fixed mock_auth_request to ensure it returns True or raises an AuthzError as expected.